ledger_device_sdk/src/io_new.rs

```diff
@@ -0,0 +1,650 @@
+        }
+        let first_len_byte = apdu_buffer[4];
+
+        match (first_len_byte, rx_len) {
+            (0, 5) => {
+                // Non-conforming zero-data APDU (TODO: per the standard, this should actually be read as a 256-byte long APDU; but that's likely to break things as lots)
+                DecodedEventType::new_apdu(header, 4, 0)
+            }
+            (0, 6) => DecodedEventType::ApduError(BadLen),
+            (0, _) => {
+                let len = u16::from_be_bytes([apdu_buffer[5], apdu_buffer[6]]) as usize;
+                if rx_len != len + 7 {
+                    return DecodedEventType::ApduError(BadLen);
+                }
+                DecodedEventType::new_apdu(header, 1 + 7, len)
+            }
+            (len, _) => {
+                if rx_len != len as usize + 5 {
+                    return DecodedEventType::ApduError(BadLen);
+                }
+                DecodedEventType::new_apdu(header, 1 + 5, len as usize)
+            }
+        }
+    }
+}
+
+/// High-level decoded event (no side effects).
+pub enum DecodedEventType {
+    Apdu {
+        header: ApduHeader,
+        offset: usize,
+        length: usize,
+    },
+    ApduError(ApduError),
+    #[cfg(not(any(target_os = "stax", target_os = "flex", target_os = "apex_p")))]
+    Button(ButtonEvent),
+    #[cfg(any(target_os = "stax", target_os = "flex", target_os = "apex_p"))]
+    Touch,
+    Ticker,
+    // Events for which no additional handling is required after decoding it
+    Ignored,
+}
+
+impl DecodedEventType {
+    fn new_apdu(header: ApduHeader, offset: usize, length: usize) -> Self {
+        Self::Apdu {
+            header,
+            offset,
+            length,
+        }
+    }
+}
+
+pub struct Command<'a, const N: usize = DEFAULT_BUF_SIZE> {
+    comm: &'a mut Comm<N>,
+    header: ApduHeader,
+    offset: usize,
+    length: usize,
+}
+
+impl<'a, const N: usize> Command<'a, N> {
+    pub fn new(comm: &'a mut Comm<N>, header: ApduHeader, offset: usize, length: usize) -> Self {
+        Self {
+            comm,
+            header,
+            offset,
+            length,
+        }
+    }
+
+    pub fn decode<T>(&self) -> Result<T, Reply>
+    where
+        T: TryFrom<ApduHeader>,
+        Reply: From<<T as TryFrom<ApduHeader>>::Error>,
+    {
+        T::try_from(self.header).map_err(Reply::from)
+    }
+
+    pub fn get_data(&self) -> &[u8] {
+        &self.comm.buf[self.offset..self.offset + self.length]
+    }
+
+    pub fn into_tx(self) -> Tx<'a, N> {
+        Tx {
+            comm: self.comm,
+            len: 0,
+        }
+    }
+
+    pub fn into_comm(self) -> &'a mut Comm<N> {
+        self.comm
+    }
+
+    pub fn reply<T: Into<Reply>>(self, data: &[u8], reply: T) -> Result<(), CommError> {
+        self.into_tx().extend(data)?.send(reply)?;
+        Ok(())
+    }
+}
+
+/// Immutable read view.
+pub struct Rx<'a, const N: usize = DEFAULT_BUF_SIZE> {
```

Comment:
Do we really need to make `Rx` public ? I mean, from the application point of view, it can be dangerous to manipulate the Rx/Tx object...